ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -366,25 +472,46 @@ IoRegisterPlugPlayNotification(IN IO_NOTIFICATION_EVENT_CATEGORY EventCategory,
 /*
  * @implemented
  */
+CODE_SEG("PAGE")
 NTSTATUS
 NTAPI
-IoUnregisterPlugPlayNotification(IN PVOID NotificationEntry)
+IoUnregisterPlugPlayNotification(
+    _In_ PVOID NotificationEntry)
 {
-    PPNP_NOTIFY_ENTRY Entry;
+    PPNP_NOTIFY_ENTRY Entry = NotificationEntry;
+    PKGUARDED_MUTEX Lock;
+
     PAGED_CODE();
 
-    Entry = (PPNP_NOTIFY_ENTRY)NotificationEntry;
     DPRINT("%s(NotificationEntry %p) called\n", __FUNCTION__, Entry);
 
-    KeAcquireGuardedMutex(&PnpNotifyListLock);
-    RemoveEntryList(&Entry->PnpNotifyList);
-    KeReleaseGuardedMutex(&PnpNotifyListLock);
-
-    RtlFreeUnicodeString(&Entry->Guid);
-
-    ObDereferenceObject(Entry->DriverObject);
+    switch (Entry->EventCategory)
+    {
+        case EventCategoryDeviceInterfaceChange:
+            Lock = &PiNotifyDeviceInterfaceLock;
+            break;
+        case EventCategoryHardwareProfileChange:
+            Lock = &PiNotifyHwProfileLock;
+            break;
+        case EventCategoryTargetDeviceChange:
+            Lock = &PiNotifyTargetDeviceLock;
+            break;
+        default:
+            UNREACHABLE;
+            return STATUS_NOT_SUPPORTED;
+    }
 
-    ExFreePoolWithTag(Entry, TAG_PNP_NOTIFY);
+    KeAcquireGuardedMutex(Lock);
+    if (!Entry->Deleted)
+    {
+        Entry->Deleted = TRUE; // so it can't be unregistered two times
```

Comment:
Does Windows protect against this, too? Otherwise I wouldn't add extra code just to support incorrect behavior.

Reply:
Yeah, it has the protection